packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -1160,6 +1190,56 @@ async function getGCStateFromSnapshot(
     return rootGCState;
 }
 
+function isMatchingGCData(
+    oldGCData?: IGarbageCollectionData,
+    newGCData?: IGarbageCollectionData,
+    newReferences?: Map<string, string[]>,
+    unreferencedTracker?: Map<string, UnreferencedStateTracker>,
+    currentReferenceTimestampMs?: number,
+): boolean {
+    // If we have any new references, gc data is not matching
+    if (oldGCData === undefined || newGCData === undefined) {
+        return false;
+    }
+
+    if (newReferences !== undefined) {
+        for (const [_, routes] of newReferences) {
+            for (const route of routes) {
+                if (newGCData.gcNodes[route] === undefined || oldGCData.gcNodes[route] === undefined) {
+                    return false;
+                }
+            }
+        }
+    }
+
+    // If we need to sweep any nodes, gc data is not matching
+    if (unreferencedTracker !== undefined && currentReferenceTimestampMs !== undefined) {
+        const unreferencedNodes = unreferencedTracker.entries();
+        for (const [_, nodeTracker] of unreferencedNodes) {
+            if (nodeTracker.shouldSweep) {
+                return false;
+            }
+        }
+    }
+
+    // Compare oldGCData and newGCData
+    const gcDataEntries1: [string, string[]][] = Object.entries(oldGCData.gcNodes);
+    for (const [nodeId1, routes1] of gcDataEntries1) {
+        const routes2: string[] | undefined = newGCData.gcNodes[nodeId1];
+        if (routes2 === undefined || routes1.length !== routes2.length) {
+            return false;
+        }
+
+        for (const route of routes1) {
+            if (!routes2.includes(route)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
```

Comment:
Any better ideas on this implementation? This is rather expensive in my mind.

Reply:
Can we do binary comparison? At the end of the day, we serialize data as strings, so there should be a way to do string comparison only, and make it (as much as possible) not GC specific.
I think the issue right now is that place where we call this function and place that actually creates blobs are separates in space and time, that causes these problems (and I think other problems, including correctness)

Reply:
The other option is to deserialize the blob, and then reserialize the blob sorted